### PR TITLE
[SPARK-44220][SQL] Move StringConcat to sql/api

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Move StringConcat to `sql/api` module.

### Why are the changes needed?

StringConcat is widely used in data types. As we plan to move entire data type family to sql/api, we should move StringConcat.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell @cloud-fan

File: sql/api/pom.xml [hunk 1]
```diff
@@ -35,6 +35,16 @@
     </properties>
 
     <dependencies>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-unsafe_${scala.binary.version}</artifactId>
```
Comment (hvanhovell) [inline sql/api/pom.xml:40]: Why do we need unsafe here?

Comment (hvanhovell) [reply sql/api/pom.xml:40]: Can you put `ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH` in common/util?

Comment (amaliujia) [reply sql/api/pom.xml:40]: good point. Yeah let me try so that we do not add too much deps to `sql/api`.

Comment (amaliujia) [reply sql/api/pom.xml]: done

**Review by cloud-fan — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks @amaliujia @cloud-fan @hvanhovell , merged to master

**Comment by panbingkun** [comment]
@amaliujia @yaooqinn @cloud-fan @hvanhovell 
From a code perspective, do we need to move the UT of this method `StringConcat` to this module? If necessary, let me do it, okay?

**Comment by hvanhovell** [comment]
@panbingkun if moving the tests is not a major hassle, then go for it!

**Comment by panbingkun** [comment]
> if moving the tests is not a major hassle, then go for it!

Ok, Got it.

**Comment by amaliujia** [comment]
@panbingkun thanks for taking care of the tests!

**Pull request closed without merging**